includes/Modules/Tag_Manager.php

```diff
@@ -85,45 +85,96 @@ final class Tag_Manager extends Module
 	public function register() {
 		$this->register_scopes_hook();
 
-		add_action( // For non-AMP.
-			'wp_head',
+		// Tag Manager tag placement logic.
+		add_action(
+			'template_redirect',
 			function() {
-				$this->print_gtm_js();
-			}
-		);
+				// Bail early if we are checking for the tag presence from the back end.
+				if ( $this->context->input()->filter( INPUT_GET, 'tagverify', FILTER_VALIDATE_BOOLEAN ) ) {
+					return;
+				}
 
-		$print_gtm_no_js = function () {
-			$this->print_gtm_no_js();
-		};
+				if ( ! $this->get_settings()->get()['useSnippet'] ) {
+					return;
+				}
 
-		// For non-AMP. WP >=5.2.
-		add_action( 'wp_body_open', $print_gtm_no_js, -9999 );
-		// For non-AMP.
-		add_action( 'wp_footer', $print_gtm_no_js );
+				// Container needs to be checked based on whether AMP or non-AMP.
+				$container_id = $this->get_data(
+					'container-id',
+					array(
+						'usageContext' => $this->context->is_amp() ? self::USAGE_CONTEXT_AMP : self::USAGE_CONTEXT_WEB,
+					)
+				);
+				if ( is_wp_error( $container_id ) || ! $container_id ) {
+					return;
+				}
 
-		$print_amp_gtm = function() {
-			// This hook is only available in AMP plugin version >=1.3, so if it
-			// has already completed, do nothing.
-			if ( ! doing_action( 'amp_print_analytics' ) && did_action( 'amp_print_analytics' ) ) {
-				return;
-			}
+				// At this point, we know the tag should be rendered, so let's take care of it
+				// for AMP and non-AMP.
+				if ( $this->context->is_amp() ) {
+					$print_amp_gtm = function() use ( $container_id ) {
+						// This hook is only available in AMP plugin version >=1.3, so if it
+						// has already completed, do nothing.
```

Comment:
Same here, this callback is used for multiple actions so this comment isn't entirely accurate.